File: android/core/src/main/java/com/clipsync/core/SessionProtocol.kt
package com.clipsync.core

import java.nio.ByteBuffer

/** CSP1 extension, negotiated by HELLO cs7|<origin>|<replay>. No clipboard is persisted. */
object SessionProtocol {
    const val PREFIX = "cs7|"
    const val MAX_TEXT = FrameCodec.MAX - 266
    fun hello(latest: ClipMessage?, replay: Boolean) = ClipMessage(MessageType.HELLO,
        lamport = latest?.lamport ?: 0, deviceId = "$PREFIX${latest?.deviceId ?: ""}|${if (replay) 1 else 0}")
    fun origin(hello: ClipMessage): String? = hello.deviceId?.takeIf { it.startsWith(PREFIX) }?.split('|')?.takeIf { it.size == 3 }?.get(1)
    fun replay(hello: ClipMessage) = hello.deviceId?.endsWith("|1") == true
    fun newer(a: ClipMessage, clock: Long, origin: String) = a.lamport > clock || (a.lamport == clock && (a.deviceId ?: "") > origin)
    fun encodeState(m: ClipMessage): ByteArray {
        val id = requireNotNull(m.deviceId).toByteArray(Charsets.UTF_8)
        val text = (m.text ?: "").toByteArray(Charsets.UTF_8)
        require(m.lamport in 1 until (Long.MAX_VALUE - 1) && id.size in 1..256 && text.size <= MAX_TEXT)
        return ByteBuffer.allocate(10 + id.size + text.size).putLong(m.lamport).putShort(id.size.toShort()).put(id).put(text).array()
    }
    fun decodeState(p: ByteArray): ClipMessage {
        require(p.size >= 11)
        val b = ByteBuffer.wrap(p); val clock = b.long; val size = b.short.toInt() and 65535
        require(clock in 1 until (Long.MAX_VALUE - 1) && size in 1..256 && size <= b.remaining())
        val id = ByteArray(size).also { b.get(it) }.toString(Charsets.UTF_8)
        val text = ByteArray(b.remaining()).also { b.get(it) }.toString(Charsets.UTF_8)
        require(text.toByteArray().size <= MAX_TEXT)
        return ClipMessage(MessageType.STATE, text = text, lamport = clock, deviceId = id)
    }
}

/** Single memory-only newest slot. Clock is logical, ties use ordinal device ID, not wall time. */
class ReconnectJournal(private val deviceId: String) {
    private var clock = 0L
    @Volatile var latest: ClipMessage? = null; private set
    @Synchronized fun local(text: String): ClipMessage {
        require(text.toByteArray().size <= SessionProtocol.MAX_TEXT)
        check(clock < Long.MAX_VALUE - 1)
        return ClipMessage(MessageType.STATE, text = text, lamport = ++clock, deviceId = deviceId).also { latest = it }
    }
    @Synchronized fun accept(message: ClipMessage): Boolean {
        require(message.lamport in 1 until (Long.MAX_VALUE - 1))
        check(clock < Long.MAX_VALUE - 1)
        clock = maxOf(clock, message.lamport) + 1
        val current = latest
        if (current != null && !SessionProtocol.newer(message, current.lamport, current.deviceId ?: "")) return false
        latest = message
        return true
    }
    @Synchronized fun observe(value: Long) { require(value in 0 until (Long.MAX_VALUE - 1)); check(clock < Long.MAX_VALUE - 1); clock = maxOf(clock, value) + 1 }
    @Synchronized fun forgetIf(message: ClipMessage) { if (latest == message) latest = null }
    @Synchronized fun clear() { latest = null }
}

File: windows/ClipSync.Core/SessionProtocol.cs
using System.Buffers.Binary;
using System.Text;
namespace ClipSync.Core;

// CSP1 extension negotiated by HELLO. Legacy TEXT sessions remain compatible.
public static class SessionProtocol
{
    public const int MaxText = FrameCodec.MaxPayloadLength - 266;
    public static ClipMessage Hello(ClipMessage? latest, bool replay) => new(MessageType.Hello,
        Lamport: latest?.Lamport ?? 0, DeviceId: $"cs7|{latest?.DeviceId ?? ""}|{(replay ? 1 : 0)}");
    public static string? Origin(ClipMessage m) { var p = m.DeviceId?.Split('|'); return p is { Length: 3 } && p[0] == "cs7" ? p[1] : null; }
    public static bool Replay(ClipMessage m) => m.DeviceId?.EndsWith("|1", StringComparison.Ordinal) == true;
    public static bool Newer(ClipMessage m, long clock, string origin) => m.Lamport > clock || (m.Lamport == clock && string.CompareOrdinal(m.DeviceId, origin) > 0);
    public static byte[] EncodeState(ClipMessage m)
    {
        var id = Encoding.UTF8.GetBytes(m.DeviceId ?? ""); var text = Encoding.UTF8.GetBytes(m.Text ?? "");
        if (m.Lamport <= 0 || m.Lamport >= long.MaxValue - 1 || id.Length is < 1 or > 256 || text.Length > MaxText) throw new InvalidDataException("Invalid state");
        var p = new byte[10 + id.Length + text.Length]; BinaryPrimitives.WriteInt64BigEndian(p, m.Lamport);
        BinaryPrimitives.WriteUInt16BigEndian(p.AsSpan(8), (ushort)id.Length); id.CopyTo(p, 10); text.CopyTo(p, 10 + id.Length); return p;
    }
    public static ClipMessage DecodeState(byte[] p)
    {
        if (p.Length < 11) throw new InvalidDataException("Short state");
        var clock = BinaryPrimitives.ReadInt64BigEndian(p); var size = BinaryPrimitives.ReadUInt16BigEndian(p.AsSpan(8));
        if (clock <= 0 || clock >= long.MaxValue - 1 || size is < 1 or > 256 || p.Length < 10 + size || p.Length - 10 - size > MaxText) throw new InvalidDataException("Invalid state");
        return new(MessageType.State, Text: Encoding.UTF8.GetString(p.AsSpan(10 + size)), Lamport: clock, DeviceId: Encoding.UTF8.GetString(p.AsSpan(10, size)));
    }
}
public sealed class ReconnectJournal(string deviceId)
{
    long _clock;
    public ClipMessage? Latest { get; private set; }
    public ClipMessage Local(string text)
    {
        if (Encoding.UTF8.GetByteCount(text) > SessionProtocol.MaxText || _clock >= long.MaxValue - 1) throw new InvalidDataException("Clipboard limit");
        return Latest = new(MessageType.State, Text: text, Lamport: ++_clock, DeviceId: deviceId);
    }
    public bool Accept(ClipMessage m)
    {
        if (m.Lamport <= 0 || m.Lamport >= long.MaxValue - 1) throw new InvalidDataException("Invalid clock");
        if (_clock >= long.MaxValue - 1) throw new InvalidDataException("Clock exhausted");
        _clock = Math.Max(_clock, m.Lamport) + 1;
        if (Latest is not null && !SessionProtocol.Newer(m, Latest.Lamport, Latest.DeviceId ?? "")) return false;
        Latest = m; return true;
    }
    public void Observe(long value) { if(value < 0 || value >= long.MaxValue - 1 || _clock >= long.MaxValue - 1) throw new InvalidDataException("Invalid clock"); _clock = Math.Max(_clock, value) + 1; }
    public void Clear() => Latest = null;
}
